### PR TITLE
Fix broken meta function for flex-attention backwards

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146563

# Summary

Fixes https://github.com/pytorch/pytorch/issues/146377

So what was the original problem: we were codegening a really weird epilogue:

```Python
        # first compute broadcasted dk of shape [Bq, Hkv, KV_LEN, V_HEAD_DIM]
        # then reduce to dk of shape [Bkv, Hkv, KV_LEN, V_HEAD_DIM]
        xindex = index_k + 64*index_n + 64*off_hkv*ks2 + 128*off_zq*ks2
        tl.store(out_ptr0 + (tl.broadcast_to(index_k + 64*index_n + off_hkv*ks1, dk.shape)), dk, mask)
        x5 = (xindex % ks3)
        tmp2 = tl.load(out_ptr0 + (x5 + ks1*off_hkv), mask, eviction_policy='evict_last')
        tl.store(out_ptr1 + (tl.broadcast_to(xindex, dk.shape)), tmp2, mask)
 ```
 
 This epilogue was writing and then reading from overlapping regions of memory causing a race condition.
 
 ### Why were we generating this epilgoue
 
 During the lowering we created a buffer w/ a different size/stride from the expected return strides. I :think this added an implicit node (for doing the permutation of this wrongly strided output to the the expected one from the meta func. The scheduler for some reason thought it was okay to fuse this into the epilogue, tbh I dont know why. 
 
 
 This fixes the broken meta func and the original repro. I will add a test but it is hard to pop, better than nothing
 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov @Chillee @yanboliang @BoyuanFeng

### CODE REVIEW

File: torch/_higher_order_ops/flex_attention.py [hunk 1]
```diff
@@ -1146,17 +1146,47 @@ def flex_attention_backward_fake_tensor_mode(
...
-        grad_query = torch.empty_like(query)
-        grad_key = torch.empty_like(key)
-        grad_value = torch.empty_like(value)
+        Bq = query.size(0)
+        Bkv = key.size(0)
+        Hkv = key.size(1)
+        seq_len_kv = key.size(2)
+        qk_head_dim = key.size(3)
+
+        grad_query = torch.empty_strided(
```
Comment (Chillee) [inline torch/_higher_order_ops/flex_attention.py:1155]: What's the difference between this and empty_like? In particular, this can output a non-contiguous buffer?

File: torch/_higher_order_ops/flex_attention.py [hunk 2]
```diff
@@ -1146,17 +1146,47 @@ def flex_attention_backward_fake_tensor_mode(
...
         )
+
+        broadcasted_grad_key = torch.empty_strided(
+            (Bq, Hkv, seq_len_kv, qk_head_dim),
+            key.stride(),
+            device=key.device,
+            dtype=key.dtype,
+        )
+
+        broadcasted_grad_value = torch.empty_strided(
```
Comment (Chillee) [inline torch/_higher_order_ops/flex_attention.py:1176]: Did we previously just not support this at all in the meta function?

**Review by Chillee — APPROVED** [review]

File: torch/_higher_order_ops/flex_attention.py (hunk 2 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:1176]: Correct, the shape would be the same though but strides could be different

File: torch/_higher_order_ops/flex_attention.py (hunk 1 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:1155]: Exactly honestly, empty like semantics are weird and the memory format is a confusing args but tldr the strides not not match exactly

Comment (bdhirsh) [reply torch/_higher_order_ops/flex_attention.py:1155]: 🤔 if you do `torch.empty_like(query, memory_format=torch.preserve_format)` you get different strides?

It seems like with preserve_format (the default) we're supposed to desugar into empty_strided anyway: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorFactories.cpp#L463

Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:1155]: Thanks Brian! I will use that for Grad Q

File: test/inductor/test_flex_attention.py [hunk 3]
```diff
@@ -2507,6 +2507,28 @@ def mask_mod(b, h, q, kv):
...
+        out_compiled = func(Q, K_sliced, V_sliced)
+
+        grad = torch.rand_like(out_eager)
+
+        eager_grads = torch.autograd.grad(out_eager, (Q, K, V), grad)
+        compiled_grads = torch.autograd.grad(out_compiled, (Q, K, V), grad)
+
+        for eager, compiled in zip(eager_grads, compiled_grads):
+            torch.testing.assert_close(eager, compiled, atol=9e-3, rtol=0)
+
```
Comment (leijurv) [inline test/inductor/test_flex_attention.py:2531]: This doesn't seem to hit the bad case for me, this test seems to already pass:

```
import torch
from torch.nn.attention.flex_attention import flex_attention

shape = (1, 2, 4096, 64)
Q = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
K = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
V = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
func = torch.compile(flex_attention, dynamic=True, fullgraph=True)

K_sliced = K[:, :, :-128]
V_sliced = V[:, :, :-128]

out_eager = flex_attention(Q, K_sliced, V_sliced)
out_compiled = func(Q, K_sliced, V_sliced)

grad = torch.rand_like(out_eager)

eager_grads = torch.autograd.grad(out_eager, (K,), grad)
compiled_grads = torch.autograd.grad(out_compiled, (K,), grad)

print(torch.isclose(eager_grads[0], compiled_grads[0], atol=9e-3, rtol=0).all())
```

File: torch/_higher_order_ops/flex_attention.py [hunk 4]
```diff
@@ -1146,17 +1146,47 @@ def flex_attention_backward_fake_tensor_mode(
     torch.Tensor, torch.Tensor, torch.Tensor, tuple[Optional[torch.Tensor], ...]
 ]:
     with mode:
-        grad_query = torch.empty_like(query)
```
Comment (leijurv) [inline torch/_higher_order_ops/flex_attention.py:1149]: This looks like https://github.com/pytorch/pytorch/commit/a3e0d4bf0797f862479c2f38d90cebd151449418, does that one also need to be empty_strided?

File: test/inductor/test_flex_attention.py (hunk 3 above)
Comment (drisspg) [reply test/inductor/test_flex_attention.py:2531]: I mentioned this in the readme but since its a race its a proxy test but not perfect, what if you run the full thing in a loop?

Comment (drisspg) [reply test/inductor/test_flex_attention.py:2531]:
```
import torch
from torch.nn.attention.flex_attention import flex_attention

torch.set_default_device("cuda")
func = torch.compile(flex_attention, dynamic=True, fullgraph=True)

K_grad = None
for i in range(10):
    torch.manual_seed(0)
    shape = (1, 2, 4096, 64)
    Q = torch.randn(shape, requires_grad=True)
    K = torch.randn(shape, requires_grad=True)
    V = torch.randn(shape, requires_grad=True)

    K_sliced = K[:, :, :-128]
    V_sliced = V[:, :, :-128]
    out = func(Q, K_sliced, V_sliced)
    print(out)
    out.sum().backward()

    if K_grad is not None:
        torch.testing.assert_close(K.grad, K_grad)
    else:
        K_grad = K.grad

```

File: torch/_higher_order_ops/flex_attention.py (hunk 4 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:1149]: Good call out!

File: torch/_higher_order_ops/flex_attention.py [hunk 5]
```diff
@@ -781,15 +781,19 @@ def sdpa_dense_backward(
     from torch._dynamo._trace_wrapped_higher_order_op import TransformGetItemToIndex
 
     # Get outputs before calling repeat interleave
-    actual_grad_query = torch.empty_like(query)
-    actual_grad_key = torch.empty_like(key)
-    actual_grad_value = torch.empty_like(value)
+    actual_grad_query = torch.empty_like(query, memory_format=torch.preserve_format)
+    actual_grad_key = torch.empty_like(key, memory_format=torch.preserve_format)
```
Comment (drisspg) [inline torch/_higher_order_ops/flex_attention.py:785]: Okay I did more tinkering and actually this should be empty_strided.

empty_like only works for the non-overlapping case but doesn't help w/ the sliced case.
Or even better we figure out the right inductor lowering to mimix empty like: https://github.com/pytorch/pytorch/blob/5f0901e57341eb9865102c1caa3d986a0c4ae3bd/torch/_inductor/kernel/flex_attention.py#L2320-L2344

However we need to have a larger tensor .. okay actualy the code I linked above just doesn't make sense I will re-thinking this tomorrow. But suffice it to say I still the the main problem is that our meta func is wrong

File: torch/_higher_order_ops/flex_attention.py [hunk 6]
```diff
@@ -1146,17 +1150,44 @@ def flex_attention_backward_fake_tensor_mode(
...
+            dtype=key.dtype,
+        )
+
+        broadcasted_grad_value = torch.empty_strided(
+            (Bq, *value.size()[1:]),
+            value.stride(),
+            device=value.device,
+            dtype=value.dtype,
+        )
+
```
Comment (janeyx99) [inline torch/_higher_order_ops/flex_attention.py:1183]: For my own understanding, is the below the multi query case

File: torch/_higher_order_ops/flex_attention.py (hunk 5 above)
Comment (zou3519) [reply torch/_higher_order_ops/flex_attention.py:785]: @drisspg the default for empty_like is preserve_format, did you change these for clarity?

File: torch/_higher_order_ops/flex_attention.py (hunk 6 above)
Comment (drisspg) [reply torch/_higher_order_ops/flex_attention.py:1184]: This isn't multiquery its just for broadcasting along Bastchdim which is useful for paged attention

**Comment by eellison** [comment]
> I think this added an implicit node (for doing the permutation of this wrongly strided output to the the expected one from the meta func. The scheduler for some reason thought it was okay to fuse this into the epilogue, tbh I dont know why.

I dont know if there is an expected behavior with wrongly strided meta functions since we are bound to get incorrect results. But generally, it is valid for inductor to fuse in a stride permutation into the epilogue (can save global mem just like other arithmetic epilogues). 

We should still look into the other bug here which is that we're writing to global memory for the output value, and then reading from global memory to compute the epilogue. This defeats the purpose of epilogue fusion. In normal template lowering, we should be computing the epilogue from the local value. Which occurs [here](https://github.com/pytorch/pytorch/blob/3008368b126f81e183a0dd157e880030b28bf77d/torch/_inductor/codegen/common.py#L2411-L2427). I don't know enough about flex attention mechanisms as to why that is not working.

**Comment by leijurv** [comment]
This feels strange. Why is the fix inside flex attention? If I'm understanding right, this should affect any triton kernel? Isn't it a super common operation to do `grad_foo = torch.empty_like(foo)` then launch your backward? It seems to me like every single instance of this is a footgun since the strides may or may not match?

File: test/inductor/test_flex_attention.py (hunk 3 above)
Comment (leijurv) [reply test/inductor/test_flex_attention.py:2531]: This version that you just sent https://github.com/pytorch/pytorch/pull/146563#discussion_r1944162178 does indeed fail for me, and adding the `.clone()` makes it succeed 👍 

However, if I add `, dtype=torch.bfloat16` to the Q K V, it no longer reproduces (it runs cleanly).

File: test/inductor/test_flex_attention.py [hunk 7]
```diff
@@ -2507,6 +2507,28 @@ def mask_mod(b, h, q, kv):
         out = func(query, key, value, block_mask=block_mask)
         out.sum().backward()
 
+    @supported_platform
+    def test_strided_backwards(self):
+        shape = (1, 2, 4096, 64)
+        Q = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
```
Comment (leijurv) [inline test/inductor/test_flex_attention.py:2513]: Possibly remove `, dtype=torch.bfloat16` per https://github.com/pytorch/pytorch/pull/146563#discussion_r1945623130

Comment (leijurv) [reply test/inductor/test_flex_attention.py:2513]: I can confirm that if you remove the bfloat16 dtype, then, this test fails on current torch for me. If you keep the bfloat16 dtype, it succeeds.

<details><summary>This version prints False</summary>

```py
import torch
from torch.nn.attention.flex_attention import flex_attention

shape = (1, 2, 4096, 64)
Q = torch.randn(shape, requires_grad=True, device="cuda")
K = torch.randn(shape, requires_grad=True, device="cuda")
V = torch.randn(shape, requires_grad=True, device="cuda")
func = torch.compile(flex_attention, dynamic=True, fullgraph=True)

K_sliced = K[:, :, :-128]
V_sliced = V[:, :, :-128]

out_eager = flex_attention(Q, K_sliced, V_sliced)
out_compiled = func(Q, K_sliced, V_sliced)

grad = torch.rand_like(out_eager)

eager_grads = torch.autograd.grad(out_eager, (K,), grad)
compiled_grads = torch.autograd.grad(out_compiled, (K,), grad)

print(torch.isclose(eager_grads[0], compiled_grads[0], atol=9e-3, rtol=0).all())
```

</details>

<details><summary>This version prints True</summary>

```py
import torch
from torch.nn.attention.flex_attention import flex_attention

shape = (1, 2, 4096, 64)
Q = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
K = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
V = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
func = torch.compile(flex_attention, dynamic=True, fullgraph=True)

K_sliced = K[:, :, :-128]
V_sliced = V[:, :, :-128]

out_eager = flex_attention(Q, K_sliced, V_sliced)
out_compiled = func(Q, K_sliced, V_sliced)

grad = torch.rand_like(out_eager)

eager_grads = torch.autograd.grad(out_eager, (K,), grad)
compiled_grads = torch.autograd.grad(out_compiled, (K,), grad)

print(torch.isclose(eager_grads[0], compiled_grads[0], atol=9e-3, rtol=0).all())
```

</details>

File: test/inductor/test_flex_attention.py [hunk 8]
```diff
@@ -2507,6 +2507,28 @@ def mask_mod(b, h, q, kv):
...
+        K = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
+        V = torch.randn(shape, requires_grad=True, device="cuda", dtype=torch.bfloat16)
```
Comment (leijurv) [inline test/inductor/test_flex_attention.py:2515]:
```suggestion
        Q = torch.randn(shape, requires_grad=True, device="cuda")
        K = torch.randn(shape, requires_grad=True, device="cuda")
        V = torch.randn(shape, requires_grad=True, device="cuda")
```

https://github.com/pytorch/pytorch/pull/146563/files#r1945630309

**Comment by leijurv** [comment]
@drisspg Might you change the test dtype from bfloat16 to float32 before merging perhaps? https://github.com/pytorch/pytorch/pull/146563#discussion_r1945634766

**Comment by drisspg** [comment]
@leijurv I am going to do it in a follow up since this CI run has been painful
